### PR TITLE
feat(engine): market-scan universe trading — no manual watchlist required

### PR DESCRIPTION
**Pull request by irfndi**
## What

**Market-scan universe trading: the watchlist becomes optional.** This PR stacks on #157 (the real position-mark fix — market trading needs honest marks to be profitable). The engine now continuously scans the Meteora universe, re-ranks candidates on a schedule, and trades the top pools through the existing gate chain — entry, watch and exit all managed automatically. No manual whitelist required.

## Why

The old model required a hand-curated `WATCHLIST_POOLS`. The deployment had 3 pools of the same pair, discovery was a single rotating page capped at 50 rows, and candidates needed 6 scans + 1h observation before they could trade. Bots doing hundreds of round-trips/day scan the whole market continuously and trade what passes their gates — this makes Prism that kind of engine.

## How it works

1. **Universe refresh** (`MARKET_SCAN_REFRESH_INTERVAL_MS`, default 30 min): new `adapter.discoverPoolsTopPages(pages)` fetches the top-N TVL-ranked pages (default 3 ≈ 3000 pools, covers everything above ~$50-100K TVL) in one call, with the Data API's token-safety metadata (verified / freeze-disabled / holders / symbol) attached. No rotating-page semantics, no 50-row slice.
2. **Market gate** (`engine/market-gate.ts`, pure + unit-tested): admits pools by
   - TVL ≥ `MARKET_SCAN_MIN_TVL_USD`
   - annualized fee APR ≥ `MARKET_SCAN_MIN_FEE_APR` (fees24h×365/tvl×100 — the profit engine)
   - volume turnover 0.02–50× (dead/wash-trading guard)
   - bin-step band (skip ultra-fine churners and ultra-wide diluters)
   - **IL/rug safety pre-filter**: stablecoins and SOL always pass; unverified or freeze-enabled legs need a real holder base; metadata absent fails open (the per-pool safety screen — blacklist/freeze/token-risk overlay — still gates every ENTER)
   - ranked by `feeAprPct × liquidityFactor`
3. **Active set** (`MARKET_SCAN_TOP_K` ≤ `MARKET_SCAN_MAX_POOLS`): rebuilt every refresh; `rebuildPoolsToScan()` assembles the per-cycle scan set = watchlist ∪ market top-K ∪ eligible candidates ∪ held positions — pools that stop qualifying drop out, **exits never stall**.
4. **Execution**: market pools skip candidate eligibility and flow straight into the per-pool loop where the FULL gate chain runs before any ENTER (safety → metrics → pre-filter → fee/IL → volume auth → bin utilization → risk → allocation). The ENTER "unmanaged pool" guard, idle-redeploy pass and per-cycle reconcile all accept `marketScanPools`.
5. **Observability**: when `autonomousTokenMode != off`, ranked pools also feed the existing `token_candidates` lifecycle.

## Config

`MARKET_SCAN_ENABLED` (off by default), `MARKET_SCAN_REFRESH_INTERVAL_MS`, `MARKET_SCAN_UNIVERSE_PAGES`, `MARKET_SCAN_MIN_TVL_USD`, `MARKET_SCAN_MIN_FEE_APR`, `MARKET_SCAN_TOP_K`, `MARKET_SCAN_MAX_POOLS`, `MARKET_SCAN_MIN_HOLDERS`, `MARKET_SCAN_MIN_BIN_STEP`, `MARKET_SCAN_MAX_BIN_STEP` — all documented in AGENTS.md.

For high-frequency operation: `SCAN_INTERVAL_MS=60000` + higher `MAX_OPEN_POSITIONS` + `MARKET_SCAN_TOP_K` in the tens → decision cycles every minute over a fresh ranked set (hundreds of round-trips/day once capital supports it).

## Tests

- `bench/market-gate.test.ts`: ranking by fee APR, TVL/fee/turnover/bin-step rejections, IL-safety pre-filter (unverified low-holder legs, freeze-enabled legs, verified-freeze trust case, stablecoin/SOL pass-through, fail-open metadata)
- config parse + clamping
- integration test: **EMPTY watchlist** + mocked universe → the gated pool is scanned and ENTERed, the rejected pool is never traded

1493 tests pass; lint, format, engine + CLI builds clean.

## Summary by Sourcery

Enable universe-driven market-scan trading that continuously ranks Meteora pools by fee APR and safety, making the manual watchlist optional while still routing all trades through existing gates.

New Features:
- Introduce a market-scan universe refresh that fetches top TVL-ranked Meteora pools with token-safety metadata and feeds them into a market gate.
- Add a pure market gate module that filters and ranks discovered pools by TVL, fee APR, volume turnover, bin step, and token safety, producing a top-K active trading set.
- Allow the engine to trade market-scan pools without manual watchlist entries while keeping held positions and candidates in the active scan set so exits and lifecycle logic continue to run.
- Expose a new adapter API to retrieve multiple pages of the Meteora pool universe in one call for market scanning.

Enhancements:
- Refactor pool discovery mapping into a shared DiscoveredPool converter that attaches token-safety metadata for both legacy discovery and market scan paths.
- Extend scan-cycle orchestration to rebuild the active pool set from watchlist, market-scan top-K, autonomous candidates, and held positions each cycle.
- Relax unmanaged-pool and idle-redeploy guards to recognize market-scan pools as managed for ENTER and redeploy decisions.

Documentation:
- Document market-scan mode, its behavior, and configuration flags in AGENTS.md, including environment variables for tuning universe refresh and gating thresholds.

Tests:
- Add unit tests for the market gate to cover ranking, filter thresholds, and token-safety pre-filter behavior.
- Add configuration tests to verify parsing, defaults, and clamping of market-scan-related env vars.
- Add an integration-style benchmark test demonstrating that market-scan mode trades from an empty watchlist while correctly excluding gated-out pools.
- Update helper config defaults to include sensible market-scan parameters used across tests.

## Summary by CodeRabbit

- **New Features**
  - Added market scanning to discover, filter, rank, and actively monitor qualifying pools.
  - Added configurable scanning controls, including refresh timing, pool limits, liquidity, fee, holder, and bin-step thresholds.
  - Added token safety checks and clearer rejection details for unsuitable pools.
  - Market scanning now integrates with candidate selection, entry decisions, redeployment, and scheduled reconciliation.
  - Scanning gracefully retains the previous pool ranking when discovery fails or returns no results.

- **Documentation**
  - Added guidance covering market-scan configuration, qualification, ranking, execution safeguards, and scan cadence.

- **Tests**
  - Added coverage for configuration handling, pool gating and ranking, and end-to-end market scanning.

### CODE REVIEW

File: engine/market-gate.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  },
+  minHolders: number,
+): boolean {
+  if (leg.isStableOrSol) return true;
+  if (leg.verified === true) {
+    return leg.freezeDisabled === false ? (leg.holders ?? 0) >= minHolders : true;
+  }
+  // Unverified leg: need freeze disabled and a real holder base.
+  return leg.freezeDisabled !== false && (leg.holders ?? 0) >= minHolders;
+}
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/market-gate.ts:86]: Fixed in ae40afa: `marketLegPasses` now treats an undefined `holders` as UNKNOWN and passes (both the unverified path and the verified+freeze-enabled path), while KNOWN holder counts keep their threshold checks — an absent count is never coerced to 0. The `bare` fixture now uses non-stable legs with NO token-safety fields so it genuinely exercises the absent-metadata path, plus a direct `marketLegPasses` unit test for absent vs known-low holder counts. 1500 tests pass; tsc/oxlint/oxfmt clean.

File: engine/adapter-service.ts [hunk 2]
```diff
@@ -3364,27 +3399,66 @@ export const AdapterLive = Layer.effect(
...
+          const baseUrl =
+            config.meteoraPoolsUrl ||
+            "https://dlmm.datapi.meteora.ag/pools?page=1&page_size=1000&filter_by=is_blacklisted=false&sort_by=tvl:desc";
+          const pageCount = Math.min(Math.max(Math.floor(pages), 1), 10);
+          const fetchPage = (page: number): Effect.Effect<ReadonlyArray<DiscoveredPool>, unknown> =>
+            Effect.gen(function* () {
+              const url = new URL(baseUrl);
+              url.searchParams.set("page", String(page));
+              url.searchParams.set("page_size", "1000");
+              const res = yield* Effect.tryPromise({
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:3422]: Fixed in ae40afa: `fetchPage` now pipes through `Effect.catchAll` that logs a per-page warning and yields `[]` for that page only — a network/timeout/parse failure on one page no longer fails `Effect.all` and discards every successfully fetched page. The outer catchAll remains as a belt-and-suspenders guard. 1500 tests pass.

File: bench/market-gate.test.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        { isStableOrSol: false, verified: true, freezeDisabled: true, holders: 10 },
+        config.minHolders,
+      ),
+    ).toBe(true);
+  });
+
+  it("fails open when token metadata is absent (per-pool screen still gates)", () => {
+    // Metadata-less pool (legacy mapping): no pre-filter veto, ranked on
+    // economics alone.
+    const bare: DiscoveredPool = { ...makePool({ address: "bare" }) };
```
[the remark this answers is not in the document]
Comment (irfndi) [reply bench/market-gate.test.ts]: Fixed in ae40afa: the `bare` fixture now omits ALL token-safety fields AND uses non-stable legs (`BareTokenX/Y` mints) — the previous fixture passed trivially via the SOL/USDC stable-leg exemptions, so it never exercised the absent-metadata path. It now genuinely asserts fail-open: absent verified/freezeDisabled/holders passes the gate, while a known low holder count still fails closed (new unit test).
